### PR TITLE
inject our own marks into the browser to identify start and end of the benchmarks

### PR DESCRIPTION
**Pull request by andersy005**
@maxrjones / @katamartin, 

as we discussed earlier today, this PR adds functionality for injecting our own marks into the browser. I'm calling these `benchmark:start` and `benchmark:end`. As you can see, we are able to measure the overall runtime (shown as `benchmark` in the timeline). 

The `benchmark:end` identifies when we observe the `onIdle` callback is invoked. 

https://github.com/carbonplan/benchmark-maps/assets/13301940/ecbd4718-bb4f-49be-b553-4e84c1f5b816

### CODE REVIEW

**Comment by katamartin** [comment]
Love how neatly the `performance.mark` events integrate into the metrics panel!

From the screen recording, it looks like the the map had not finished rendering by the time `benchmark:end` was marked. Is that right? If so, could this have to do with the units/offsetting in the different time measures? Or is the idle callback simply getting invoked too early?

**Comment by maxrjones** [comment]
> From the screen recording, it looks like the the map had not finished rendering by the time benchmark:end was marked. Is that right? If so, could this have to do with the units/offsetting in the different time measures? Or is the idle callback simply getting invoked too early?

It would be worthwhile if @andersy005 could confirm this, but I think this is an artifact of how the screenshots are rendered in the performance UI. I see the same thing zoomed out for some runs, but when I zoom in the first screenshot after `benchmark:end` shows a completed map.

Why do you think `.zmetadata` is requested before the start of the benchmark?

**Comment by katamartin** [comment]
> Why do you think `.zmetadata` is requested before the start of the benchmark?

Good question! Based on the way `maps` + the prototype are implemented, we should at least be confident that `.zmetadata` will not be requested until `Display` button is clicked. Looking at `main.py`, it seems like there might be a meaningful delay between these two events: 
```
  page.click('//div[text()="Display"]/following-sibling::button')

  page.evaluate(
      """
      () => (window.performance.mark("benchmark:start"))
                """
  )
```

Reading the Node.js docs for the [Page class](https://playwright.dev/docs/api/class-page#page-click), it looks like these are both natively async methods. Because we're using the [`playwright.sync_api`](https://playwright.dev/python/docs/api/class-page), it seems likely that these methods are both getting automatically `await`ed, which would be responsible for the delay we're seeing. I think we could either swap the order of these events (and risk capturing the time between `benchmark:start` and activating the `Display` button) and/or switch to the `playright.async_api` and explicitly `await` the actions we want to occur in sequence and omit in cases where we don't (like this combination of events).

**Comment by andersy005** [comment]
> From the screen recording, it looks like the the map had not finished rendering by the time benchmark:end was marked. Is that right? If so, could this have to do with the units/offsetting in the different time measures? Or is the idle callback simply getting invoked too early?

i just recorded two runs, and i'm noticing different behaviors:

in the first screenshot, the `idle` callback seems to have been invoked early.  


Screenshot 2023-07-26 at 15 17 29

----

for the this run, i'm seeing a finished map. 

Screenshot 2023-07-26 at 15 16 55

**Comment by andersy005** [comment]
> I think we could either swap the order of these events (and risk capturing the time between benchmark:start and activating the Display button)


here are the results of this swap:

Screenshot 2023-07-26 at 15 26 07

---
---

> and/or switch to the playright.async_api and explicitly await the actions we want to occur in sequence and omit in cases where we don't (like this combination of events).

👍🏽  i'm happy to look 👀 into this...

**Comment by katamartin** [comment]
After letting that digest for a little bit, I think that switching to `playright.async_api` and confirming which behavior we want for each of invocations of `playright` methods feels like the safer path. I think in that world, we should also probably still swap the two actions we're talking about for good measure.

**Comment by katamartin** [comment]
>i just recorded two runs, and i'm noticing different behaviors:

Are these behaviors confirmed when you "zoom in" on the screenshots in the trace like @maxrjones alluded to?

File: main.py [hunk 1]
```diff
@@ -50,30 +51,30 @@ def run(
...
+    await page.goto(url)
 
     # Focus on and click the map element
-    page.focus('.mapboxgl-canvas')
-    page.click('.mapboxgl-canvas')
+    await page.focus('.mapboxgl-canvas')
+    await page.click('.mapboxgl-canvas')
     # use performance.mark API to mark the start of the benchmarks.
-    page.evaluate(
+    await page.evaluate(
```
Comment (katamartin) [inline main.py]: This is a case where we'd want to fire this off synchronously, right?
```suggestion
    page.evaluate(
```

File: main.py [hunk 2]
```diff
@@ -50,30 +51,38 @@ def run(
...
+                  """
+    )
 
     # click the button that is a sibling of the div with the text "Display".
-    page.click('//div[text()="Display"]/following-sibling::button')
+    await page.click('//div[text()="Display"]/following-sibling::button')
 
     # Start timer
-    page.evaluate(
+    await page.evaluate(
```
Comment (katamartin) [inline main.py]: This seems like another case where we might not want to use `async`

File: main.py (hunk 1 above)
Comment (andersy005) [reply main.py]: good catch. for these two operations to be executed at the same time, the following should work

```python
    await asyncio.gather(
        page.evaluate(
            """
            () => (window.performance.mark("benchmark:start"))
                      """
        ),
        page.click('//div[text()="Display"]/following-sibling::button')
    )

```

**Comment by andersy005** [comment]
here's the latest run with `async` API:

https://github.com/carbonplan/benchmark-maps/assets/13301940/74b5bf5f-3750-4dcb-89b7-7e26d00b2304

**Pull request merged**